Blog Post: Advancements in Lightning Infrastructure (#432)
OpenSats/website: data/blog/advancements-in-lightning-infrastructure.mdx

censorship resistance to ensure Bitcoin remains a tool for financial freedom.

In addition to supporting Dusty's work on Splicing, OpenSats has also funded an
initiative to integrate [Splicing into
LDK](/blog/bitcoin-grants-feb-2024#splicing-in-ldk). Over the past several
months, efforts to integrate Splicing into LDK have yielded meaningful strides
that led to a key breakthrough: the integration of [Interactive Transaction
Construction](https://github.com/lightningdevkit/rust-lightning/pull/2419). This
feature closely ties in with [Dual
Funding](https://github.com/lightningdevkit/rust-lightning/issues/1621#issue-1308076236),
a mechanism that allows two participants to collaboratively open channels.
Because Splicing and Dual Funding share underlying transaction logic, progress
on one directly benefits the other.

This work has required frequent releases to keep pace with LDK's rapid evolution
and ensure compatibility with the latest protocol updates. In parallel, Splicing
messages (including `splice_locked`) have been implemented, along with support
for `Pay-After-Splice` and `Splice-After-Pay`—features that allow channels to

> be resized dynamically without interrupting active transactions.

Reviewer: This paragraph should be dropped completely, as it contains minute details that don't add much info in this context.

> be resized dynamically without interrupting active transactions.

Author: ```suggestion
```

As part of the ongoing effort to integrate Splicing into LDK, project developer
[Optout](https://github.com/optout21) has been conducting end-to-end testing in

> `ldk-node`, `ldk-sample`, and `ldk-node-sample`, assessing stability and

Reviewer: Remove `ldk-node` from the list, as it is not relevant

> `ldk-node`, `ldk-sample`, and `ldk-node-sample`, assessing stability and

Author: ```suggestion
`ldk-sample`, and `ldk-node-sample`, assessing stability and
```

> readiness for integration. Although several foundational pull requests have been
> merged, further work is needed to refine interoperability and finalize message
> flows before Splicing is fully production-ready within LDK. Once complete, it

Reviewer: ```suggestion
readiness for integration. Although several foundational pull requests have been
merged, further work is needed to finalize integration into LDK and
ensure interoperability before Splicing is fully production-ready within LDK. Once complete, it
```

project has achieved a significant milestone by contributing to the integration
of LDK with `lnprototest`, a test suite designed to rigorously validate
Lightning protocol behavior. This work enhances LDK's reliability, ensures
adherence to Lightning standards, and strengthens its overall robustness, giving
developers greater confidence when testing their implementations. A key part of
this effort was the launch of the
[`ldk-lnprototest`](https://pypi.org/project/ldk-lnprototest/) package on PyPI.
This package provides a runner that facilitates the integration of LDK with
`lnprototest`, enabling automated and standardized testing. By reducing the
risks of protocol inconsistencies and implementation errors—challenges that can
significantly impact Lightning's reliability—this tool ensures a more robust and
predictable development process for LDK-based implementations like Lampo.

>Building reliable Lightning infrastructure requires both rigorous testing and
>continuous improvements. The LDK-Lnprototest integration and enhancements to
>Lampo aim to make the network more robust and developer-friendly. Excited to
>push this forward! 

> >—Prakhar Saxena, Lampo Project Lead

Reviewer 2: ```suggestion
>—Prakhar Saxena, Lampo Project Lead
```

Note: Just to point out that I am the author of lampo and the current only maintainer of the project, I am happy to leave Prakhar as a project lead 😄

In addition to improving the Lightning testing environment, Lampo has also
[introduced
improvements](https://github.com/vincenzopalazzo/lampo.rs/pull/295/commits) to
wallet security and usability. Previously, users had to manually enter their
[BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) seed words in the
command line every time they restored a wallet, posing privacy risks and
reducing usability. Now, Lampo securely stores and manages wallet restoration
data in a binary file (`wallet.dat`), eliminating the need for repeated manual
entry and improving the user experience without compromising security.

Together, these improvements make Lampo both a reliable development tool and a
more secure node implementation.

> Looking ahead, Lampo plans to continue its evolution by implementing a
> proof-of-concept for Validating Lightning Signer (VLS), a move that will further
> improve security by enabling external validation of Lightning signatures. Future
> updates will also introduce
> [BOLT12](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md#bolt-12-negotiation-protocol-for-lightning-payments)
> support, refine the plugin system, and expand documentation to encourage broader
> adoption. By fostering a truly open-source, modular approach to running a

Reviewer 2: ```suggestion
Looking ahead, Lampo plans to continue its evolution by implementing a
proof-of-concept for Validating Lightning Signer (VLS), a move that will further
improve security by enabling external validation of Lightning signatures. Future
updates will also introduce a PoC of a plugin system, and expand documentation to encourage broader
adoption. By fostering a truly open-source, modular approach to running a
```

Lampo supported BOLT 12 before ocean released their support, you can see lampo as one of the first lightning network implementations that support the recent version of bolt12 thanks to ldk

to specific implementations, reinforcing its role as a vital infrastructure
component for the Lightning Network.

With OpenSats
[support](/blog/bitcoin-grants-july-2024#lightning-network-protocol-test-framework),
one important advancement made has been the introduction of out-of-tree testing,
which allows tests to reside within an implementation's codebase rather than
being confined to the lnprototest repository. This not only reduces clutter but
also improves scalability, enabling projects to maintain their test suites while
still leveraging the standardized framework. Additionally, the development of
custom runners—such as LampoRunner—demonstrates how implementations can
tailor the testing process to their specific needs, further improving
flexibility and usability. These refinements make lnprototest a more powerful
tool for ensuring protocol compliance and catching issues before they impact
users.

By providing a stable and modular testing environment, lnprototest plays a
crucial role in safeguarding the reliability of the Lightning Network. Without a
rigorous testing framework, protocol changes risk breaking interoperability

> between implementations, leading to fragmentation and unreliable payment flows.

Reviewer 2: ```suggestion
between implementations, leading to fragmentation.
```

payment flow usually are not related to protocol bugs per-se

>make payments. This is why it is so important to have independent developers
>building testing tools and researching the health of the protocol as a whole.

Looking ahead, BLAST is preparing for the next development phase, focusing on
creating lightweight models for different LN implementations, followed by
automation features that enable reproducible network states. Incremental
improvements will continue rolling out, ensuring that BLAST remains a meaningful
tool for interoperability research. As Lightning adoption grows, the need for
independent, large-scale testing frameworks will only become more critical, and
BLAST is helping to lay the groundwork for a more well-structured Lightning
Network.

### Lampo

As the Lightning Network evolves, the need for modular, open-source node
implementations has become increasingly clear. Existing implementations often
come with trade-offs that limit flexibility for developers seeking highly
customizable solutions. Addressing this gap, [Lampo](https://lampo.devcrew.cc/)
provides a modular, community-driven alternative, built on the Lightning
Development Kit (LDK).

With [support](/blog/8th-wave-of-bitcoin-grants#lampo) from OpenSats, the

> project has achieved a significant milestone by integrating with `lnprototest`,

Author: ```suggestion
project has achieved a significant milestone by contributing to the integration of LDK with `lnprototest`,
```

a test suite designed to rigorously validate Lightning protocol behavior. This

> work enhances Lampo's reliability, ensures adherence to Lightning standards, and

Author: ```suggestion
work enhances LDK's reliability, ensures adherence to Lightning standards, and
```

> strengthens LDK's overall robustness, giving developers greater confidence when

Author: ```suggestion
strengthens its overall robustness, giving developers greater confidence when
```

> testing their implementations. By enabling automated, standardized testing

Author: ```suggestion
testing their implementations. A key part of this effort was the launch of the [`ldk-lnprototest`](https://pypi.org/project/ldk-lnprototest/) package on PyPI. This package provides a runner that facilitates the integration of LDK with `lnprototest`, enabling automated and standardized testing. By enabling automated, standardized testing
```

> through `lnprototest`, Lampo helps reduce the risks of protocol inconsistencies
> and implementation errors—challenges that can significantly impact Lightning's
> reliability.

Author: ```suggestion
```
